docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,354 @@
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new typescript.Client(
+  name: "@acme/infra-client",
+);
+
+// The deployer of the client host will need to retrieve/link this data later
+let clientContext = new cloud.Bucket();
+clientContext.addObject("ctx", client.context);
+```
+
+When compiling to non-simulator targets, the client will store the expected permissions to be used elsewhere.
+For example, in `tf-aws`, the client will generate a policy to be attached to a role or user. The name of the policy is required to be set in the `wing.toml` file.
+
+```toml
+[tf-aws]
+client_policy = "MyClientPolicy"
+```
+
+After the client is generated and published, it can be used in any TypeScript project:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// Will use the environment variable WING_CONTEXT to resolve the context
+const client = new Client();
+
+// Optionally, you can pass the context directly to bypass the environment variable
+// const client = new Client({ context: "..." });
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+It's read through the environment variable `WING_CONTEXT` and is expected to be a base64 encoded.
+
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions for the simulator
+
+The client is effectively an `IInflightHost`, so it any resource lifted into it registers the permissions it needs.
+This is implemented on a per-target basis to expose the permissions in a useful way.
+For the `sim` target this problem is (currently) much more difficult.
+The simulator is centralized and will have no knowledge of the future clients that will be using it.
+The simulator has no API to add new permissions after the simulator has started and the scope of that feels too large for this RFC.
+
+As an MVP, the simulator can be configured to ignore permissions via `wing.toml`.
+Even though this is unsafe and cause cause local/cloud differences, wing is intentionally not claiming ownership of the client's host environment.
+
+### Type generation with `@lift`
+
+Wing's compiler has type information available for use, but only before preflight execution occurs.
+This makes an API like the following not feasible to implement using logic similar to the existing `extern` or `@inflight`:
+
+```wing
+client.lift("bucket", bucket, ["put"]);
```

Comment:
Why not use lift expressions here?

```suggestion
client.lift(@lift({ bucket: ["put"] });
```

Reply:
This section describes why we can't use the former syntax